File: C Code/pr7.c
#include <stdio.h>
int main() {
    int n, quantum;
    printf("Enter number of processes: ");
    scanf("%d", &n);
    int process_id[n], burst_time[n], arrival_time[n];
    int remaining_time[n], waiting_time[n], turnaround_time[n], completion_time[n];
    int total_waiting = 0, total_turnaround = 0;
    for (int i = 0; i < n; i++) {
        process_id[i] = i + 1;
        printf("Enter Burst Time for P%d: ", i + 1);
        scanf("%d", &burst_time[i]);
        printf("Enter Arrival Time for P%d: ", i + 1);
        scanf("%d", &arrival_time[i]);
        remaining_time[i] = burst_time[i];
        waiting_time[i] = 0;
        turnaround_time[i] = 0;
    }
    printf("Enter Time Quantum: ");
    scanf("%d", &quantum);
    int time = 0, done;
    do {
        done = 1;
        for (int i = 0; i < n; i++) {
            if (remaining_time[i] > 0 && arrival_time[i] <= time) {
                done = 0;
                if (remaining_time[i] > quantum) {
                    time += quantum;
                    remaining_time[i] -= quantum;
                } else {
                    time += remaining_time[i];
                    waiting_time[i] = time - burst_time[i] - arrival_time[i];
                    remaining_time[i] = 0;
                    completion_time[i] = time;
                }
            } else if (arrival_time[i] > time) {
                time++;
            }
        }
    } while (!done);
    for (int i = 0; i < n; i++) {
        turnaround_time[i] = burst_time[i] + waiting_time[i];
        total_waiting += waiting_time[i];
        total_turnaround += turnaround_time[i];
    }
    printf("\nProcess\tAT\tBT\tCT\tWT\tTAT\n");
    for (int i = 0; i < n; i++) {
        printf("P%d\t%d\t%d\t%d\t%d\t%d\n",
               process_id[i], arrival_time[i], burst_time[i],
               completion_time[i], waiting_time[i], turnaround_time[i]);
    }
    float avg_waiting = (float)total_waiting / n;
    float avg_turnaround = (float)total_turnaround / n;
    printf("\nAverage Waiting Time: %.2f\n", avg_waiting);
    printf("Average Turnaround Time: %.2f\n", avg_turnaround);
    return 0;
}

File: pr7.py
#include<stdio.h>
#include<stdlib.h>
#include<string.h>
int main(){
int n,q,i,j,time=0,finished_count=0;
printf("Number of processes: ");
scanf("%d",&n);
int burst[n],remaining[n],waiting[n],turnaround[n],finished[n];
printf("Enter %d burst times (space separated): ",n);
for(i=0;i<n;i++){
scanf("%d",&burst[i]);
remaining[i]=burst[i];
waiting[i]=0;
finished[i]=0;
}
printf("Time quantum: ");
scanf("%d",&q);
while(finished_count<n){
int made_progress=0;
for(i=0;i<n;i++){
if(remaining[i]>0){
made_progress=1;
if(remaining[i]>q){
time+=q;
remaining[i]-=q;
}else{
time+=remaining[i];
waiting[i]=time-burst[i];
remaining[i]=0;
finished[i]=1;
finished_count++;
}
}
}
if(!made_progress)break;
}
for(i=0;i<n;i++)
turnaround[i]=waiting[i]+burst[i];
float avg_wait=0,avg_turnaround=0;
for(i=0;i<n;i++){
avg_wait+=waiting[i];
avg_turnaround+=turnaround[i];
}
avg_wait/=n;
avg_turnaround/=n;
printf("\nPID\tBurst\tWaiting\tTurnaround\n");
for(i=0;i<n;i++)
printf("P%d\t%d\t%d\t%d\n",i+1,burst[i],waiting[i],turnaround[i]);
printf("\nAverage waiting time: %.2f\n",avg_wait);
printf("Average turnaround time: %.2f\n",avg_turnaround);
return 0;
}